src/login/loginCoordinator.ts

```diff
@@ -0,0 +1,286 @@
+	token?: string;
+}
+
+interface LoginOptions {
+	safeHostname: string;
+	url: string | undefined;
+	autoLogin?: boolean;
+}
+
+/**
+ * Coordinates login prompts across windows and prevents duplicate dialogs.
+ */
+export class LoginCoordinator {
+	private readonly inProgressLogins = new Map<string, Promise<LoginResult>>();
+
+	constructor(
+		private readonly secretsManager: SecretsManager,
+		private readonly mementoManager: MementoManager,
+		private readonly vscodeProposed: typeof vscode,
+		private readonly logger: Logger,
+	) {}
+
+	/**
+	 * Direct login - for user-initiated login via commands.
+	 * Stores session auth and URL history on success.
+	 */
+	public async promptForLogin(
+		options: LoginOptions & { url: string },
+	): Promise<LoginResult> {
+		const { safeHostname, url } = options;
+		return this.executeWithGuard(safeHostname, async () => {
+			const result = await this.attemptLogin(
+				{ safeHostname, url },
+				options.autoLogin ?? false,
+			);
+
+			await this.persistSessionAuth(result, safeHostname, url);
+
+			return result;
+		});
+	}
+
+	/**
+	 * Shows dialog then login - for system-initiated auth (remote).
+	 */
+	public async promptForLoginWithDialog(
+		options: LoginOptions & { message?: string; detailPrefix?: string },
+	): Promise<LoginResult> {
+		const { safeHostname, url, detailPrefix, message } = options;
+		return this.executeWithGuard(safeHostname, () => {
+			// Show dialog promise
+			const dialogPromise = this.vscodeProposed.window
+				.showErrorMessage(
+					message || "Authentication Required",
+					{
+						modal: true,
+						useCustom: true,
+						detail:
+							(detailPrefix || `Authentication needed for ${safeHostname}.`) +
+							"\n\nIf you've already logged in, you may close this dialog.",
+					},
+					"Login",
+				)
+				.then(async (action) => {
+					if (action === "Login") {
+						// Proceed with the login flow, handling logging in from another window
+						const storedAuth =
+							await this.secretsManager.getSessionAuth(safeHostname);
+						const newUrl = await maybeAskUrl(
+							this.mementoManager,
+							url,
+							storedAuth?.url,
+						);
+						if (!newUrl) {
+							throw new Error("URL must be provided");
+						}
+
+						const result = await this.attemptLogin(
+							{ url: newUrl, safeHostname },
+							false,
+						);
+
+						await this.persistSessionAuth(result, safeHostname, newUrl);
+
+						return result;
+					} else {
+						// User cancelled
+						return { success: false };
+					}
+				});
+
+			// Race between user clicking login and cross-window detection
+			return Promise.race([
+				dialogPromise,
+				this.waitForCrossWindowLogin(safeHostname),
+			]);
+		});
+	}
+
+	private async persistSessionAuth(
+		result: LoginResult,
+		safeHostname: string,
+		url: string,
+	): Promise<void> {
+		// Empty token is valid for mTLS
+		if (result.success && result.token !== undefined) {
+			await this.secretsManager.setSessionAuth(safeHostname, {
```

Comment:
Is it safe to skip calling this for mTLS?  That would cause `onDidChangeSessionAuth` to never fire right?  And then that could possibly cause a number of things waiting for auth to never see it.